core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -2004,6 +2020,131 @@ private[spark] class DAGScheduler(
     }
   }
 
+  /**
+   * Schedules shuffle merge finalize.
+   */
+  private[scheduler] def scheduleShuffleMergeFinalize(stage: ShuffleMapStage): Unit = {
+    logInfo(("%s (%s) scheduled for finalizing" +
+      " shuffle merge in %s s").format(stage, stage.name, shuffleMergeFinalizeWaitSec))
+    shuffleMergeFinalizeScheduler.schedule(
+      new Runnable {
+        override def run(): Unit = finalizeShuffleMerge(stage)
+      },
+      shuffleMergeFinalizeWaitSec,
+      TimeUnit.SECONDS
+    )
+  }
+
+  /**
+   * DAGScheduler notifies all the remote shuffle services chosen to serve shuffle merge request for
+   * the given shuffle map stage to finalize the shuffle merge process for this shuffle. This is
+   * invoked in a separate thread to reduce the impact on the DAGScheduler main thread, as the
+   * scheduler might need to talk to 1000s of shuffle services to finalize shuffle merge.
+   */
+  private[scheduler] def finalizeShuffleMerge(stage: ShuffleMapStage): Unit = {
+    logInfo("%s (%s) finalizing the shuffle merge".format(stage, stage.name))
```

Comment:
What happens if the stage was cancelled during `shuffleMergeFinalizeWaitSec` ?

Reply:
Addressed this comment after discussing with @mridulm offline. Mridul, I tried adding test for the cancellation but the `DAGSchedulerEventProcessLoopTester` is not async as it simply forwards the event for immediate processing. Let me know if you have other ideas to test this particular situation.

Reply:
Added additional tests to handle the cases of stage cancellation, barrier stage, late arrival of merge results etc.